src/stapi_fastapi/models/order.py

```diff
@@ -1,15 +1,44 @@
-from typing import Literal
+from typing import Generic, Literal, TypeVar
 
 from geojson_pydantic import Feature, FeatureCollection
 from geojson_pydantic.geometries import Geometry
 from pydantic import BaseModel, ConfigDict, Field, StrictInt, StrictStr
 
+from stapi_fastapi.models.opportunity import OpportunityProperties
 from stapi_fastapi.models.shared import Link
 from stapi_fastapi.types.datetime_interval import DatetimeInterval
+from stapi_fastapi.types.filter import CQL2Filter
 
 
-class OrderProperties(BaseModel):
+class OrderParameters(BaseModel):
+    model_config = ConfigDict(extra="allow")
+
+
+ORP = TypeVar("ORP", bound=OrderParameters)
+OPP = TypeVar("OPP", bound=OpportunityProperties)
+
+
+class OrderRequest(BaseModel, Generic[ORP]):
+    datetime: DatetimeInterval
+    geometry: Geometry
+    # TODO: validate the CQL2 filter?
+    filter: CQL2Filter | None = None
+
+    order_parameters: ORP
+
+    model_config = ConfigDict(strict=True)
+
+
+class OrderProperties(BaseModel, Generic[OPP, ORP]):
```

Comment:
I see what you did here, makes complete sense. Except `geometry`, that's already reflected in the `Order` object as it is a `Feature`.

I do think we don't have `Order` or `OrderCollection` typed appropriately; In both cases we need to make allow passing the type for `OrderProperties`, I think:

```python
class Order(Feature[Geometry, OrderProperties[OPP, ORP]]):
# and
class OrderCollection(FeatureCollection[Order[OPP, ORP]]):
```

At least that's the pattern from the opportunities models signatures. (I'd have put comments on each, but they fall outside the changeset and thus can't have comments).

The other thing is that we had specifically set `ConfigDict(extra="forbid")` on `OrderParameters` to raise an error if the requestor includes some parameters that are not recognized. For an endpoint that could charge people money it seemed reasonable to be as strict as possible as to not inadvertently let someone thing something they requested was honored when it was simply not recognized.

But in doing this it meant that the generic top-level `/orders` endpoints wouldn't be able to return orders because they wouldn't have the specific order model per product to do the correct validation. Thus we made `order_parameters: dict[str, Any]` to support anything. Doing that ensures that `OrderProperties`, which is only used for orders in responses, has no tie to `OrderParameters`, which we want to be used only for requests.

Oh, and I'm not sure it makes sense to support a null `filter`. That is part of the request model to make specifying the filter optional. But I think here we can set it to an empty filter `{}`? Of course if this works then I think they type might be wrong everywhere; requests should be defaulted to `{}` instead of `None.

Reply:
Fixed most of these.

For CQL2, `{}` isn't a valid CQL2 JSON entity, so we need to use `null` as the default/undefined value.